REGRESSION (278815@main): [ MacOS iOS Debug ] TestWebKitAPI.AdvancedPrivacyProtections.AddNoiseToWebAudioAPIs is a consistent failure

whsieh opened the pull request:
#### 1e5a82de122e187fb66f39af69abc2a35557b72c
<pre>
REGRESSION (278815@main): [ MacOS iOS Debug ] TestWebKitAPI.AdvancedPrivacyProtections.AddNoiseToWebAudioAPIs is a consistent failure
<a href="https://bugs.webkit.org/show_bug.cgi?id=274266">https://bugs.webkit.org/show_bug.cgi?id=274266</a>
<a href="https://rdar.apple.com/128209071">rdar://128209071</a>

Reviewed by Chris Dumez.

Allocations are forbidden here for performance reasons:

```
void AudioDestinationNode::renderQuantum(…)
{
    …

    // For performance reasons, we forbid heap allocations while doing rendering on the audio thread.
    // Heap allocations that cannot be avoided or have not been fixed yet can be allowed using
    // DisableMallocRestrictionsForCurrentThreadScope scope variables.
    ForbidMallocUseForCurrentThreadScope forbidMallocUse;
```

When applying noise (in the case where Advanced Privacy Protections are enabled), honor this by
stack-allocating the noise multiplier buffer instead.

* Source/WebCore/platform/audio/AudioUtilities.cpp:
(WebCore::AudioUtilities::applyNoise):

Canonical link: <a href="https://commits.webkit.org/278945@main">https://commits.webkit.org/278945@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/52d35bf7554d44a483f84325d11a359547be906c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/52055 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/31381 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/4464 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/55328 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/2767 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/54357 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/37762 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/2473 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/42367 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/1766 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/54149 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/28998 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/44902 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/23438 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/26278 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/2176 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/937 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/48176 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/2325 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/56925 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/27170 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/2314 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [⏳ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/GTK-WK2-Tests-EWS "Waiting to run tests") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/28408 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/45028 "Passed tests") | [⏳ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/API-Tests-GTK-EWS "Waiting to run tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11384 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/29311 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/28148 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

hortont424 commented on Source/WebCore/platform/audio/AudioUtilities.cpp:
```diff
@@ -80,6 +80,8 @@ size_t timeToSampleFrame(double time, double sampleRate, SampleFrameRounding rou
 
 void applyNoise(float* values, size_t numberOfElementsToProcess, float standardDeviation)
 {
+    DisableMallocRestrictionsForCurrentThreadScope disableMallocRestrictions;
```
Is this the right fix? Or should we be doing the allocations ahead of time somehow? Slowing down the audio thread in PB more than we need to just for the noise seems bad, no?

cdumez replied:
Right, it'd be better if we could avoid heap allocations (didn't check if this was a viable option).

whsieh replied:
The newest version of this patch should address this, mostly — while it'll be lazily allocated the first time, we'll keep reusing the same buffer for subsequent noise injection.

cdumez commented on Source/WebCore/platform/audio/AudioUtilities.cpp:
```diff
@@ -27,10 +27,10 @@
 #if ENABLE(WEB_AUDIO)
 
 #include "AudioUtilities.h"
+#include "ThreadGlobalData.h"
```
No longer needed.

whsieh replied:
Ah, good catch.

cdumez commented on Source/WebCore/platform/audio/AudioUtilities.cpp:
```diff
@@ -84,24 +84,21 @@ void applyNoise(float* values, size_t numberOfElementsToProcess, float standardD
-    auto computeNoiseMultiplier = [&](float rawValue) {
-        if (!noiseMultipliers.isValidKey(rawValue))
-            return distribution(generator);
+    static constexpr auto maximumTableSize = 2048;
```
`static` is not necessary here given that this is `constexpr`

cdumez commented on Source/WebCore/platform/audio/AudioUtilities.cpp:
```diff
@@ -84,24 +84,21 @@ void applyNoise(float* values, size_t numberOfElementsToProcess, float standardD
-        if (!noiseMultipliers.isValidKey(rawValue))
-            return distribution(generator);
+    static constexpr auto maximumTableSize = 2048;
+    static constexpr auto minimumTableSize = 1024;
```
ditto.

cdumez commented on Source/WebCore/platform/audio/AudioUtilities.cpp:
```diff
@@ -84,24 +84,21 @@ void applyNoise(float* values, size_t numberOfElementsToProcess, float standardD
-        auto result = noiseMultipliers.ensure(rawValue, [&] {
-            return distribution(generator);
-        }).iterator->value;
+    size_t unusedCellCount = weakRandomNumber<uint32_t>() % (maximumTableSize - minimumTableSize);
```
We may want a comment to explain that we're using this instead of a HashMap in order to avoid heap allocations on the audio rendering thread.

cdumez commented on Source/WebCore/platform/audio/AudioUtilities.cpp:
```diff
@@ -84,24 +84,21 @@ void applyNoise(float* values, size_t numberOfElementsToProcess, float standardD
-        if (noiseMultipliers.size() >= maxNoiseMultiplierMapSize)
-            noiseMultipliers.remove(noiseMultipliers.random());
+    std::array<float, maximumTableSize> multipliers;
+    std::fill(std::begin(multipliers), std::end(multipliers), std::nanf(""));
```
I think we can simplify this with:
```
multipliers.fill(std::numeric_limits<float>::quiet_NaN());
```

cdumez replied:
We may also consider using `std::numeric_limits<float>::infinity()` instead NaN for the empty value, which is what our HashTable uses by default.

whsieh replied:
Sounds good! Will go with `std::numeric_limits<float>::infinity()` and `!std::isfinite` to check.

cdumez reviewed (approved):
LGTM with nits.

cdumez commented on Source/WebCore/platform/audio/AudioUtilities.cpp:
```diff
@@ -84,24 +83,22 @@ void applyNoise(float* values, size_t numberOfElementsToProcess, float standardD
-        values[i] *= computeNoiseMultiplier(values[i]);
+    for (size_t i = 0; i < numberOfElementsToProcess; ++i) {
+        auto& multiplier = multipliers[DefaultHash<double>::hash(values[i]) % tableSize];
+        if (!std::isfinite(multiplier))
```
nit: Could use `std::isinf()`

whsieh committed:
REGRESSION (278815@main): [ MacOS iOS Debug ] TestWebKitAPI.AdvancedPrivacyProtections.AddNoiseToWebAudioAPIs is a consistent failure
https://bugs.webkit.org/show_bug.cgi?id=274266
rdar://128209071

Reviewed by Chris Dumez.

Allocations are forbidden here for performance reasons:

```
void AudioDestinationNode::renderQuantum(…)
{
    …

    // For performance reasons, we forbid heap allocations while doing rendering on the audio thread.
    // Heap allocations that cannot be avoided or have not been fixed yet can be allowed using
    // DisableMallocRestrictionsForCurrentThreadScope scope variables.
    ForbidMallocUseForCurrentThreadScope forbidMallocUse;
```

When applying noise (in the case where Advanced Privacy Protections are enabled), honor this by
stack-allocating the noise multiplier buffer instead.

* Source/WebCore/platform/audio/AudioUtilities.cpp:
(WebCore::AudioUtilities::applyNoise):

Canonical link: https://commits.webkit.org/278945@main
